### PR TITLE
Add Object Lock default retention configuration for S3 buckets

### PR DESCRIPTION
**Pull request by abraverm**
##### SUMMARY
https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-configure.html#object-lock-configure-set-retention-period-object

Design detail:
AWS API doesn't support unsetting the default retention, though it is possible in the Web console.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
s3_bucket

### CODE REVIEW

File: plugins/modules/s3_bucket.py [hunk 1]
```diff
@@ -286,6 +307,14 @@
...
+# Default Object Lock retention
+- amazon.aws.s3_bucket:
+    name: mys3bucket
+    state: present
+    object_lock_default_retention:
+      mode: governance
+      days: 1
 """
 
 RETURN = r"""
```
Comment (hakbailey) [inline plugins/modules/s3_bucket.py:320]: We should update the RETURN block documentation to include the new `"object_lock_default_retention"` value.

File: plugins/modules/s3_bucket.py [hunk 2]
```diff
@@ -166,6 +166,27 @@
     type: bool
     default: false
     version_added: 6.0.0
+  object_lock_default_retention:
+    description:
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply plugins/modules/s3_bucket.py:170]: We should probably also note here that an `object_lock_default_retention` can be updated but not removed.

**Review by hakbailey — COMMENTED** [review]
This looks good! See a few mostly minor comments below.

**Comment by abraverm** [comment]
Thank you @hakbailey, I have addressed your comments, could you give it another look?

**Comment by abraverm** [comment]
retry

**Comment by abraverm** [comment]
recheck

File: plugins/modules/s3_bucket.py [hunk 3]
```diff
@@ -832,6 +874,56 @@ def handle_bucket_object_lock(s3_client, module: AnsibleAWSModule, name: str) ->
...
+        if object_lock_enabled:
+            object_lock_configuration_status = get_object_lock_configuration(s3_client, name)
+        else:
+            object_lock_configuration_status = {}
+    except is_boto3_error_code(["NotImplemented", "XNotImplemented"]) as e:
+        if object_lock_default_retention is not None:
+            module.fail_json(msg="Fetching bucket object lock default retention is not supported")
+    except is_boto3_error_code("AccessDenied") as e:  # pylint: disable=duplicate-except
+        if object_lock_default_retention is not None:
+            module.fail_json(msg="Permission denied fetching object lock default retention for bucket")
```
Comment (hakbailey) [inline plugins/modules/s3_bucket.py:902]:
```suggestion
            module.fail_json_aws(e, msg="Permission denied fetching object lock default retention for bucket")
```

File: plugins/modules/s3_bucket.py [hunk 4]
```diff
@@ -1733,10 +1863,21 @@ def main():
         validate_bucket_name=dict(type="bool", default=True),
         dualstack=dict(default=False, type="bool"),
         object_lock_enabled=dict(type="bool"),
+        object_lock_default_retention=dict(
+            type="dict",
+            options=dict(
+                mode=dict(type="str", choices=["GOVERNANCE", ["COMPLIANCE"]], required=True),
```
Comment (hakbailey) [inline plugins/modules/s3_bucket.py:1869]:
```suggestion
                mode=dict(type="str", choices=["GOVERNANCE", "COMPLIANCE"], required=True),
```

**Review by hakbailey — COMMENTED** [review]
A few more minor comments. In addition, running `tox -e black` locally should fix the linting errors.

**Review by alinabuzachis — COMMENTED** [review]
@abraverm Thank you for working on this feature idea. Can you please add a changelog fragment (https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) for this change?

**Review by alinabuzachis — APPROVED** [review]

**Review by hakbailey — APPROVED** [review]

**Comment by abraverm** [comment]
Thank you very much for merging this PR, I was a bit busy and couldn't respond on time.

**Pull request merged**